akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -738,38 +778,52 @@ private[akka] class ShardRegion(
   }
 
   def replyToRegionStateQuery(ref: ActorRef): Unit = {
-    askAllShards[Shard.CurrentShardState](Shard.GetCurrentShardState)
-      .map { shardStates =>
-        CurrentShardRegionState(shardStates.map {
-          case (shardId, state) => ShardRegion.ShardState(shardId, state.entityIds)
-        }.toSet)
-      }
-      .recover {
-        case _: AskTimeoutException => CurrentShardRegionState(Set.empty)
+    queryShards[Shard.CurrentShardState](shards, Shard.GetCurrentShardState)
+      .map { qr =>
+        // Productionize CurrentShardRegionState #27406
+        val state =
+          qr.responses.map(state => ShardRegion.ShardState(state.shardId, state.entityIds)) ++
+          qr.failed.map(sid => ShardRegion.ShardState(sid, Set.empty))
+        CurrentShardRegionState(state.toSet)
       }
       .pipeTo(ref)
   }
 
   def replyToRegionStatsQuery(ref: ActorRef): Unit = {
-    askAllShards[Shard.ShardStats](Shard.GetShardStats)
-      .map { shardStats =>
-        ShardRegionStats(shardStats.map {
-          case (shardId, stats) => (shardId, stats.entityCount)
-        }.toMap)
-      }
-      .recover {
-        case _: AskTimeoutException => ShardRegionStats(Map.empty)
+    queryShards[ShardStats](shards, Shard.GetShardStats)
+      .map { qr =>
+        ShardRegionStats(qr.responses.map(stats => (stats.shardId, stats.entityCount)).toMap, qr.failed)
       }
       .pipeTo(ref)
   }
 
-  def askAllShards[T: ClassTag](msg: Any): Future[Seq[(ShardId, T)]] = {
+  /**
+   * Query all or a subset of shards, e.g. unresponsive shards that initially timed out.
+   * If the number of `shards` are less than this.shards.size, this could be a retry.
+   * Returns a partitioned set of any shards that may have not replied within the
+   * timeout and shards that did reply, to provide retry on only that subset.
+   *
+   * Logs a warning if any of the group timed out.
+   *
+   * To check subset unresponsive: {{{ queryShards[T](shards.filterKeys(u.contains), shardQuery) }}}
+   */
+  def queryShards[T: ClassTag](shards: Map[ShardId, ActorRef], msg: Any): Future[ShardsQueryResult[T]] = {
     implicit val timeout: Timeout = settings.shardRegionQueryTimeout
-    Future.sequence(shards.toSeq.map {
-      case (shardId, ref) => (ref ? msg).mapTo[T].map(t => (shardId, t))
-    })
+
+    Future.traverse(shards.toSeq) { case (shardId, shard) => askOne(shard, msg, shardId) }.map { ps =>
+      val qr = ShardsQueryResult[T](ps, this.shards.size)
+      if (qr.failed.nonEmpty) log.warning(qr.toString)
```

Comment:
```suggestion
      if (qr.failed.nonEmpty) log.warning(s"Some shards could not be queried: [$qr]")
```

Reply:
The `ShardsQueryResult`'s `toString` has logic in it for granular reporting - this is also what is now more available future status details, and **so far** only used by logging, aside from the failed set returned in the larger query. I think it might be a bit redundant to prepend further.
```scala
override val toString: String = {
      if (total == 0)
        s"Shard region had zero shards to gather metadata from."
      else if (isTotalFailed || isAllSubsetFailed) {
        s"All [${failed.size}] shards ${if (isAllSubsetFailed) "of subset" else ""} queried were unresponsive."
      } else {
        s"Queried [$queried] shards of [$total]: responsive [${responses.size}], unresponsive [${failed.size}] within the timeout."
      }
    }
```